import os

samples_path = os.path.abspath('samples.csv')

with open(samples_path, 'r') as f:
    samples = [i.strip().split(',')[1][:-11] for i in f.readlines()]

fastqc_out = os.path.abspath("test_output/fastqc")
spades_out = os.path.abspath("test_output/spades")
quast_out = os.path.abspath("test_output/quast")

rule all:
    input:
        expand(f'{fastqc_out}/{{sample}}_1_fastqc.html', sample=samples),
        expand(f'{spades_out}/{{sample}}/scaffolds.fasta', sample=samples),
        expand(f'{spades_out}/{{sample}}/contigs.fasta', sample=samples),
        expand(f'{quast_out}/{{sample}}/report.txt', sample=samples),


# Контроль качества ридов (FastQC)
rule fastqc:
    input:
        # in_1 = './test_input/SRR31122807_1.fastq.gz',
        # in_2 = './test_input/SRR31122807_2.fastq.gz'
        in_1 = 'test_input/{sample}_1.fastq.gz',
        in_2 = 'test_input/{sample}_2.fastq.gz'

    conda: 'envs/fastqc.yaml'
    output:
        fastqc_html_1 = f'{fastqc_out}/{{sample}}_1_fastqc.html',
        fatqc_zip_1 = f'{fastqc_out}/{{sample}}_1_fastqc.zip',
        fastqc_html_2 = f'{fastqc_out}/{{sample}}_2_fastqc.html',
        fatqc_zip_2 = f'{fastqc_out}/{{sample}}_2_fastqc.zip'
    params: 
        fastqc_outdir = fastqc_out
    shell:
        'fastqc {input.in_1} {input.in_2} -o {params.fastqc_outdir}'
    

# Сборка генома (SPAdes)
rule spades:
    input:
        in_1 = 'test_input/{sample}_1.fastq.gz',
        in_2 = 'test_input/{sample}_2.fastq.gz'
    conda: "./envs/spades.yaml"
    output:
        spades_scaffolds = f'{spades_out}/{{sample}}/scaffolds.fasta',
        spades_contigs = f'{spades_out}/{{sample}}/contigs.fasta'

    params: 
        spades_outdir = f'{spades_out}/{{sample}}'
    shell:
        'spades.py -1 {input.in_1} -2 {input.in_2} -o {params.spades_outdir}'


# Оценка качества сборки (QUAST)
rule quast:
    input:
        in_1 = f'{spades_out}/{{sample}}/scaffolds.fasta',
        in_2 =  f'{spades_out}/{{sample}}/contigs.fasta'
    conda: "./envs/quast.yaml"
    output:
        quast_report = f'{quast_out}/{{sample}}/report.txt'
    params: 
        quast_outdir = f'{quast_out}/{{sample}}'
    shell:
        'quast.py {input.in_1} {input.in_2} -o {params.quast_outdir}'
